content/blog/2024-01-26-more-secrets-options.md

```diff
@@ -0,0 +1,312 @@
+
+At present, any external secret objects will need to be deleted from the repository to use
+the `none` backend - since the ArgoCD application will not sync when a non-existing CRD is
+referenced.
+
+## How to Use a non-default Backend
+
+We have provided Makefile targets to switch between backends. These targets edit the pattern
+configuration files in-place; in keeping with the GitOps philosophy they change the files in
+git that control how the application is deployed.
+
+These Makefile targets are:
+
+```text
+  secrets-backend-vault                Edits values files to use default Vault+ESO secrets config
+  secrets-backend-kubernetes           Edits values file to use Kubernetes+ESO secrets config
+  secrets-backend-none                 Edits values files to remove secrets manager + ESO
+```
+
+Run the makefile target in your repo to effect the necessary changes. If the command makes changes,
+it will display them in `git diff` format, and it will be up to you to commit and push the result
+to effect the change. Nothing will change on your cluster until you commit and push.
+
+## Using the old system - The `legacy-load-secrets` Makefile target
+
+The existing vault-utils codepath is available via the `legacy-load-secrets`
+Makefile target. If secrets loading fails, or you just want to use the other
+system, you can run `make legacy-load-secrets` after `make install` and it will
+run those scripts and the Ansible playbooks and roles associated with them.
+
+## Deprecation of v1.0 Secrets
+
+The v1.0 secrets format has not been used in the Validated Patterns framework
+for over a year now. The v2.0 framework is a strict superset of the v1.0
+framework. Support for the v1.0 framework is still available via the
+`legacy-load-secrets` code path, but this may be removed in the future.
+
+## Updates to the Secrets v2.0 Schema
+
+New features have been added at both the file level and the per-secret level
+to support the new backends:
+
+### Top-level Additions
+
+#### `secretStoreNamespace`
+
+example:
+```yaml
+---
+version: "2.0"
+secretStoreNamespace: 'validated-patterns-secrets'
+
+secrets:
+```
+
+A new top-level key has been introduced in the secrets file:
+`secretStoreNamespace`. This defaults to `validated-patterns-secrets`. This is
+the namespace that ESO uses as its special secret store, which serves the same
+architectural purpose as vault does in the default installation. (Secrets are
+installed into this namespace as Kubernetes secrets objects, and ESO allows for
+them to be copied or templated out using ESO mechanisms).
+
+#### `defaultAnnotations`
+
+example:
+
+```yaml
+defaultAnnotations:
+  validatedpatterns.io/pattern: 'myPattern'
+```
+
+This data structure is a hash, or dictionary. These labels will be applied to all
+secrets objects unless they have per-secret annotations set. Labels are only added to
+kubernetes based secrets objects (using the `kubernetes` or `none`) backends. The
```

Comment:
I think what I'm missing is why this behaviour is required and/or useful for the k8s backend but not vault.

Reply:
The main difference here is that the framework previously supported only native vault objects (in the sense that vault objects are not inherently kubernetes things). The main way to use annotations and labels directly in Kubernetes would be for the "none" backend (I don't know if they would be useful generally on secrets used as backing stores by ESO; but since in general setting labels and annotations is a basic k8s object property, common to all k8s objects, it seemed like a good idea to support it here too.)